src/herder/TxQueueLimiter.cpp

```diff
@@ -10,244 +10,177 @@
-        }
-        auto curOps = tx->getNumOperations();
-        if (neededOps <= curOps)
-        {
-            return std::make_pair(true, 0ll);
-        }
-        neededOps -= curOps;
+        resetTxs();
     }
 
-    // we reach this point if the queue doesn't have capacity for that
-    // transaction even when empty. Combination of multiplier and max ledger
-    // size is too small for whatever reason
-    static size_t lastMax = std::numeric_limits<size_t>::max();
-    if (lastMax != maxQueueSizeOps())
+    int64_t minFeeToBeatEvicted = computeBetterFee(mEvictedFeeBid, *newTx);
+    if (minFeeToBeatEvicted > 0)
     {
-        lastMax = maxQueueSizeOps();
-        CLOG_WARNING(Herder,
-                     "Transaction Queue limiter configured with {} operations: "
-                     "node won't be able to accept all transactions",
-                     lastMax);
+        return std::make_pair(false, minFeeToBeatEvicted);
     }
 
-    return std::make_pair(false, 0ll);
-}
-
-TransactionFrameBasePtr
-TxQueueLimiter::getWorstTransaction()
-{
-    auto it = mTxs->begin();
-    if (it == mTxs->end())
+    uint32_t txOpsDiscount = 0;
+    if (oldTx)
     {
-        throw std::logic_error(
-            "invalid state getting worst tx in TxQueueLimiter");
+        auto newTxOps = newTx->getNumOperations();
+        auto oldTxOps = oldTx->getNumOperations();
+        // Bump transactions must have at least the old amount of operations.
+        releaseAssert(oldTxOps <= newTxOps);
+        txOpsDiscount = newTxOps - oldTxOps;
     }
-    return *it;
+    return mTxs->canFitWithEviction(*newTx, txOpsDiscount, txsToEvict);
 }
 
-bool
+void
 TxQueueLimiter::evictTransactions(
-    size_t ops, std::function<void(TransactionFrameBasePtr const&)> evict)
+    std::vector<TxStackPtr> const& txsToEvict,
+    std::function<void(TransactionFrameBasePtr const&)> evict)
 {
-    while (size() + ops > maxQueueSizeOps())
+    for (auto const& evictedStack : txsToEvict)
     {
-        if (size() == 0)
-        {
-            return false;
-        }
-        auto evictTx = getWorstTransaction();
-        mMinFeeNeeded = {evictTx->getFeeBid(), evictTx->getNumOperations()};
-        evict(evictTx);
+        auto tx = evictedStack->getTopTx();
+        mEvictedFeeBid = {tx->getFeeBid(), tx->getNumOperations()};
+        evict(tx);
     }
-    return true;
 }
 
 void
 TxQueueLimiter::reset()
 {
-    mTxs = std::make_unique<QueueLimiterTxMap>();
-    mQueueSizeOps = 0;
-    resetMinFeeNeeded();
+    resetTxs();
+    resetEvictionState();
 }
 
-std::pair<int64, uint32>
-TxQueueLimiter::getMinFeeNeeded() const
+void
+TxQueueLimiter::resetEvictionState()
 {
-    return mMinFeeNeeded;
+    // We cannot normally initialize `mTxs` queue in the constructor
+    // because `maxQueueSizeOps()` may not be initialized. Hence we initialize
+    // lazily during the add/reset.
+    if (mTxs == nullptr)
+    {
+        resetTxs();
```

Comment:
Why do you need this? The only path where this applies is when `resetEvictionState` is called from `TransactionQueue::shift()` when `TxQueueLimiter` is empty it seems.

Reply:
TBH I don't remember why exactly did I add this, doesn't seem like removing this breaks anything.